File: benchmarks/tsb/bench_timestamp_str_format.ts
/**
 * Benchmark: Timestamp string formatting — strftime, isoformat, day_name, month_name.
 * Outputs JSON: {"function": "timestamp_str_format", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Timestamp } from "../../src/index.ts";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 50;

const timestamps = Array.from(
  { length: SIZE },
  (_, i) => new Timestamp(Date.UTC(2020, i % 12, (i % 28) + 1, i % 24, i % 60, i % 60)),
);

for (let i = 0; i < WARMUP; i++) {
  for (const ts of timestamps) {
    ts.strftime("%Y-%m-%d %H:%M:%S");
    ts.isoformat();
    ts.day_name();
    ts.month_name();
  }
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  for (const ts of timestamps) {
    ts.strftime("%Y-%m-%d %H:%M:%S");
    ts.isoformat();
    ts.day_name();
    ts.month_name();
  }
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "timestamp_str_format",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_timestamp_str_format.py
"""Benchmark: Timestamp string formatting — strftime, isoformat, day_name, month_name."""
import json
import time
import pandas as pd

SIZE = 10_000
WARMUP = 5
ITERATIONS = 50

timestamps = [
    pd.Timestamp(year=2020, month=(i % 12) + 1, day=(i % 28) + 1,
                 hour=i % 24, minute=i % 60, second=i % 60)
    for i in range(SIZE)
]

for _ in range(WARMUP):
    for ts in timestamps:
        ts.strftime("%Y-%m-%d %H:%M:%S")
        ts.isoformat()
        ts.day_name()
        ts.month_name()

start = time.perf_counter()
for _ in range(ITERATIONS):
    for ts in timestamps:
        ts.strftime("%Y-%m-%d %H:%M:%S")
        ts.isoformat()
        ts.day_name()
        ts.month_name()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "timestamp_str_format",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
